### PR TITLE
[SPARK-33141][SQL] Capture SQL configs when creating permanent views

### PR DESCRIPTION
**Pull request by luluorta**
### What changes were proposed in this pull request?
This PR makes CreateViewCommand/AlterViewAsCommand capturing runtime SQL configs and store them as view properties. These configs will be applied during the parsing and analysis phases of the view resolution. Users can set `spark.sql.legacy.useCurrentConfigsForView` to `true` to restore the behavior before.

### Why are the changes needed?
This PR is a sub-task of [SPARK-33138](https://issues.apache.org/jira/browse/SPARK-33138) that proposes to unify temp view and permanent view behaviors. This PR makes permanent views mimicking the temp view behavior that "fixes" view semantic by directly storing resolved LogicalPlan. For example, if a user uses spark 2.4 to create a view that contains null values from division-by-zero expressions, she may not want that other users' queries which reference her view throw exceptions when running on spark 3.x with ansi mode on.

### Does this PR introduce _any_ user-facing change?
No.

### How was this patch tested?
added UT + existing UTs (improved)

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala [hunk 1]
```diff
@@ -132,9 +132,11 @@ object AnalysisContext {
  */
 class Analyzer(
     override val catalogManager: CatalogManager,
-    conf: SQLConf)
+    deprecatedConf: SQLConf)
```
Comment (maropu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:135]: Is this change related to what this PR proposes?

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/AnalysisSuite.scala [hunk 2]
```diff
@@ -771,22 +771,24 @@ class AnalysisSuite extends AnalysisTest with Matchers {
...
-      $"a" / $"d" as "div4",
-      $"e" / $"e" as "div5")
-
-    val message = intercept[TreeNodeException[LogicalPlan]] {
-      testAnalyzer.execute(plan)
-    }.getMessage
-    assert(message.startsWith(s"Max iterations ($maxIterations) reached for batch Resolution, " +
-      s"please set '${SQLConf.ANALYZER_MAX_ITERATIONS.key}' to a larger value."))
+    withSQLConf(SQLConf.ANALYZER_MAX_ITERATIONS.key -> maxIterations.toString) {
+      val conf = new SQLConf().copy(SQLConf.ANALYZER_MAX_ITERATIONS -> maxIterations)
```
Comment (maropu) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/AnalysisSuite.scala:775]: ditto: https://github.com/apache/spark/pull/30289/files#r519499900

File: sql/core/src/test/scala/org/apache/spark/sql/execution/SQLViewSuite.scala [hunk 3]
```diff
@@ -748,4 +748,30 @@ abstract class SQLViewSuite extends QueryTest with SQLTestUtils {
       }
     }
   }
+
+  test("SPARK-33141 view should be parsed and analyzed with configs set when creating") {
```
Comment (maropu) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/SQLViewSuite.scala:752]: nit: `SPARK-33141` -> `SPARK-33141: `

**Comment by maropu** [comment]
Could you describe more about the `Why are the changes needed?` section? In my first impression, since the view is a logical one, the current behaivour (the behaivour of logical views depends on current runtime configs) looks more natural than the proposed one. cc: @cloud-fan @viirya

**Comment by viirya** [comment]
BTW, capturing configs into views may hit config migration issue. We might add/change/remove configs across Spark versions.

**Comment by viirya** [comment]
> Could you describe more about the `Why are the changes needed?` section? In my first impression, since the view is a logical one, the current behaivour (the behaivour of logical views depends on current runtime configs) looks more natural than the proposed one. cc: @cloud-fan @viirya

What configs we should capture? As @maropu said, a view is basically logical query, is it important to capture SQL configs for it? From the description, seems it is for keeping semantically consistency. For a logical query, will we change its semantic by using different config values? If there is, it is behavior breaking config actually. For such configs, capturing a view's configs seems not making sense too.

Can you describe the use case?

**Comment by cloud-fan** [comment]
can we copy-paste more context from JIRA tickets to the PR description? @luluorta

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/interface.scala [hunk 4]
```diff
@@ -414,6 +433,8 @@ object CatalogTable {
   val VIEW_QUERY_OUTPUT_PREFIX = VIEW_PREFIX + "query.out."
   val VIEW_QUERY_OUTPUT_NUM_COLUMNS = VIEW_QUERY_OUTPUT_PREFIX + "numCols"
   val VIEW_QUERY_OUTPUT_COLUMN_NAME_PREFIX = VIEW_QUERY_OUTPUT_PREFIX + "col."
+
+  val VIEW_QUERY_SQL_CONFIGS = VIEW_PREFIX + "query.sqlConfigs"
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/interface.scala:437]: This is similar to current catalog/namespace, which is about the context, not query. Can we define it close to `VIEW_CATALOG_AND_NAMESPACE` and follow it's property key naming?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala [hunk 5]
```diff
@@ -1448,6 +1448,15 @@ object SQLConf {
         "must be positive.")
       .createWithDefault(100)
 
+  val APPLY_VIEW_SQL_CONFIGS =
+    buildConf("spark.sql.legacy.view.applySQLConfigs")
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:1452]: legacy config name should describe the legacy behavior. How about `spark.sql.legacy.useCurrentConfigsForView`?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/views.scala [hunk 6]
```diff
@@ -334,6 +337,21 @@ case class ShowViewsCommand(
 
 object ViewHelper {
 
+  private val configPrefixBlacklist = Seq(
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/views.scala:340]: nit: `configPrefixDenyList`

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/views.scala [hunk 7]
```diff
@@ -334,6 +337,21 @@ case class ShowViewsCommand(
...
+    SQLConf.MAX_NESTED_VIEW_DEPTH.key,
+    "spark.sql.optimizer.",
+    "spark.sql.codegen.",
+    "spark.sql.execution.",
+    "spark.sql.shuffle.",
+    "spark.sql.adaptive.")
+
+  private def isConfigBlacklisted(key: String): Boolean = {
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/views.scala:348]: nit: 
```
def shouldCaptureConfig(key: String): Boolean = {
  !configPrefixDenyList.exists(prefix => key.startsWith(prefix))
}
```

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/views.scala [hunk 8]
```diff
@@ -380,16 +425,19 @@ object ViewHelper {
     // for createViewCommand queryOutput may be different from fieldNames
     val queryOutput = analyzedPlan.schema.fieldNames
 
+    val conf = SQLConf.get
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/views.scala:428]: since we have `session` passed in, seems better to use `session.sessionState.conf`

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/views.scala [hunk 9]
```diff
@@ -361,11 +379,38 @@ object ViewHelper {
     }
   }
 
+  /**
+   * Convert the view query SQL configs in `properties`.
+   */
+  private def generateQuerySQLConfigs(conf: SQLConf): Map[String, String] = {
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/views.scala:385]: `sqlConfigsToProps`, following `catalogAndNamespaceToProps`

File: sql/core/src/test/scala/org/apache/spark/sql/execution/SQLViewSuite.scala [hunk 10]
```diff
@@ -754,4 +754,77 @@ abstract class SQLViewSuite extends QueryTest with SQLTestUtils {
       }
     }
   }
+
+  test("SPARK-33141: view should be parsed and analyzed with configs set when creating") {
+    withTable("t33141") {
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/SQLViewSuite.scala:759]: it's hard to read, let's just use `t`, `v1`, `v2`, etc.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/views.scala [hunk 11]
```diff
@@ -361,11 +379,38 @@ object ViewHelper {
...
+    val modifiedConfs = conf.getAllConfs.filter { case (k, _) =>
+      conf.isModifiable(k) && !isConfigBlacklisted(k)
+    }
+    val props = new mutable.HashMap[String, String]
+    if (modifiedConfs.nonEmpty) {
+      val confJson = compact(render(JsonProtocol.mapToJson(modifiedConfs)))
+      props.put(VIEW_QUERY_SQL_CONFIGS, confJson)
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/views.scala:392]: have you stress-tested this? The hive metastore has a limitation about property value length. You can take a look at `HiveExternalCatalog.tableMetaToTableProps`.

Another idea is to put one config per table property entry.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala (hunk 1 above)
Comment (luluorta) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: This change is related to another sub-task of [SPARK-33138](https://issues.apache.org/jira/browse/SPARK-33138) which makes internal classes of SparkSession always using active SQLConf. I opened a seperated PR (https://github.com/apache/spark/pull/30299) and removed this change here.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/AnalysisSuite.scala (hunk 2 above)
Comment (luluorta) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/AnalysisSuite.scala]: ditto

File: sql/core/src/test/scala/org/apache/spark/sql/execution/SQLViewSuite.scala (hunk 3 above)
Comment (luluorta) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/SQLViewSuite.scala]: done

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/views.scala [hunk 12]
```diff
@@ -361,11 +364,34 @@ object ViewHelper {
     }
   }
 
+  /**
+   * Convert the view query SQL configs in `properties`.
+   */
+  private def generateQuerySQLConfigs(conf: SQLConf): Map[String, String] = {
+    val modifiedConfs = conf.getAllConfs.filter { case (k, _) =>
+      conf.isModifiable(k) && k != SQLConf.MAX_NESTED_VIEW_DEPTH.key
```
[the remark this answers is not in the document]
Comment (luluorta) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/command/views.scala]: IMHO, we should keep the captured SQL configs as less as possible. Because it's hard to precisely captures the configs which only affect the parser and analyzer, an alternative way is just filtering out the configs that definitely can NOT affect parsing/analyzing.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/interface.scala (hunk 4 above)
Comment (luluorta) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/interface.scala]: done

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala (hunk 5 above)
Comment (luluorta) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala]: done

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/views.scala (hunk 6 above)
Comment (luluorta) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/command/views.scala]: done

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/views.scala (hunk 7 above)
Comment (luluorta) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/command/views.scala]: done

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/views.scala (hunk 8 above)
Comment (luluorta) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/command/views.scala]: done

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/views.scala (hunk 9 above)
Comment (luluorta) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/command/views.scala]: done

File: sql/core/src/test/scala/org/apache/spark/sql/execution/SQLViewSuite.scala (hunk 10 above)
Comment (luluorta) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/SQLViewSuite.scala]: done

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/views.scala (hunk 11 above)
Comment (luluorta) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/command/views.scala]: Thanks for pointing this out. I changed to store one config per table property entry, each with a "view.sqlConfig." prefix.

**Comment by luluorta** [comment]
> can we copy-paste more context from JIRA tickets to the PR description? 

I added more background information to the PR description.

**Comment by cloud-fan** [comment]
My two cents:
1. It's unfortunate that Spark has "semantic" configs which can change query result. This is bad for a SQL system, but we can't get rid of it now.
2. For a view, it's more reasonable to "fix" its semantic after created since the semantic can change by configs. If a user turns on ANSI mode, he should still be able to read existing views that may not be valid under ANSI mode.

**Comment by cloud-fan** [comment]
thanks, merging to master!
